### PR TITLE
Skip projects not selected for build

### LINKED ISSUE #89: Skipping projects not selected for build
With the `-f X.sln` switch GitLink is smart enough to only process projects from the solution, which is great.

I've found however that it also tries to process projects not selected for building given the **Configuration**/**Platform** combination. 

Maybe they should be skipped by GitLink as well?

### PR DESCRIPTION
**Pull request by tpluscode**
Fixes #89 by ignoring projects which are not selected for building in solution's Configuration Manager.

The necessary information is available in the [`ProjectConfigurationInSolution` class](https://msdn.microsoft.com/en-us/library/microsoft.build.construction.ProjectConfigurationInSolution.aspx). Solution configurations are accessible through [this property](https://msdn.microsoft.com/en-us/library/microsoft.build.construction.projectinsolution.projectconfigurations.aspx).

I'm using ImpromptuInterface for easier access to types internal to `Microsoft.Build` while avoiding reflection and dynamic.

### CODE REVIEW

**Comment by tpluscode** [comment]
I have a small issue with `TypeCache`. In `ProjectHelper` there is the line

```
SolutionParserType = TypeCache.GetType("Microsoft.Build.Construction.SolutionParser");
```

When running the entire test suite the `SolutionParserType` is always null. When only the `ProjectHelperFacts` test fixture is run it works. Why is that?

**Comment by GeertvanHorrik** [comment]
You might need to use LoadAssembliesOnStartup. The reason is that the assembly is embedded (using Costura), but if you use stuff via reflection (string based), there is no need for the .net framework to load the assembly into memory (resulting into a null result when looking up the type).

LoadAssembliesOnStartup adds a typeof(firsttypefromeachreferencedassembly) to ensure types are loaded into the AppDomain. This allows the app to always use all the types and not having to wait for a real usage of a type to lazy-load the assembly.

A workaround to test this would be to add typeof(sometypefromMicrosoft.Build) into the ModuleInitializer of the app and see if that solves the issue.

**Comment by tpluscode** [comment]
Thanks, LoadAssembliesOnStartup does the trick

**Comment by tpluscode** [comment]
Hm, I assume you would like me to only include the single assembly for LoadAssembliesOnStartup so that nothing else is unnecessarily loaded?

```
<LoadAssembliesOnStartup>
    <IncludeAssemblies>
        Microsoft.Build
    </IncludeAssemblies>
</LoadAssembliesOnStartup>
```

**Comment by GeertvanHorrik** [comment]
@tpluscode Yep, that's the right approach.

**Comment by GeertvanHorrik** [comment]
Just ping me when it's ready for review, then I'll check it.

**Comment by tpluscode** [comment]
I just changed that LoadAssembliesOnStartup. You can have look now.

File: src/GitLink/Helpers/ProjectHelper.cs [hunk 1]
```diff
@@ -17,37 +18,37 @@ namespace GitLink
...
         private static readonly Type SolutionParserType;
-        private static readonly PropertyInfo SolutionReaderPropertyInfo;
-        private static readonly PropertyInfo ProjectsPropertyInfo;
-        private static readonly MethodInfo ParseSolutionMethodInfo;
-        private static readonly PropertyInfo RelativePathPropertyInfo;
-        private static readonly PropertyInfo ProjectTypePropertyInfo;
         private static readonly object KnownToBeMsBuildFormat;
 
-        static ProjectHelper()
+        public interface ISolutionParser
```
Comment (GeertvanHorrik) [inline src/GitLink/Helpers/ProjectHelper.cs]: Please move interfaces to separate files

**Comment by GeertvanHorrik** [comment]
Looks good. The ImpromptuInterface is a nice library if I see the usage, thanks for adding. If you move the interfaces to separate files (in the _Interfaces_ directory, then it's ready to be merged.

**Comment by tpluscode** [comment]
Okay, I moved the interfaces. I had to add that directory right?

File: src/GitLink/Helpers/ProjectHelper.cs [hunk 2]
```diff
@@ -42,7 +42,7 @@ static ProjectHelper()
         public static IEnumerable<Project> GetProjects(string solutionFile, string configurationName, string platformName)
         {
             var projects = new List<Project>();
-            var solutionParser = SolutionParserType.GetConstructors(BindingFlags.Instance | BindingFlags.NonPublic).First().Invoke(null).ActLike<ISolutionParser>();
+            var solutionParser = ((object)Impromptu.InvokeConstructor(SolutionParserType)).ActLike<ISolutionParser>();
```
Comment (tpluscode) [inline src/GitLink/Helpers/ProjectHelper.cs]: I figure this is a slightly more readable semantically

File: src/GitLink/Helpers/ProjectHelper.cs [hunk 3]
```diff
@@ -17,37 +18,19 @@ namespace GitLink
     using Microsoft.Build.Evaluation;
     using System.IO;
     using System.Text.RegularExpressions;
+    using ImpromptuInterface;
+    using Interfaces;
```
Comment (GeertvanHorrik) [inline src/GitLink/Helpers/ProjectHelper.cs]: No namespace for interfaces, just a folder so they are all together. Same for .Helpers, .Extensions, etc. A namespace should represent a functionality group, not a folder on disk.

File: src/GitLink/Interfaces/IProjectConfigurationInSolution.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+namespace GitLink.Interfaces
```
Comment (GeertvanHorrik) [inline src/GitLink/Interfaces/IProjectConfigurationInSolution.cs]: namespace GitLink

File: src/GitLink/Interfaces/IProjectInSolution.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+namespace GitLink.Interfaces
```
Comment (GeertvanHorrik) [inline src/GitLink/Interfaces/IProjectInSolution.cs]: namespace GitLink

File: src/GitLink/Interfaces/ISolutionParser.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+namespace GitLink.Interfaces
```
Comment (GeertvanHorrik) [inline src/GitLink/Interfaces/ISolutionParser.cs]: namespace GitLink

Comment (tpluscode) [inline src/GitLink/Interfaces/ISolutionParser.cs]: So I also exclude the folder with Resharper shared settings? I mean the [Namespace provider](http://stackoverflow.com/questions/6821579/where-does-resharper-store-the-namespace-provider-settings) option

**Comment by tpluscode** [comment]
And actually, how about these interfaces go to some meaningful namespace? The actual types are in the `Microsoft.Build.Construction` namespace. Maybe top level `GitLink` namespace is not the best place for them. Something like `GitLink.Build.Construction`?

**Comment by GeertvanHorrik** [comment]
Good point. I am not 100 % sure yet, but feel free to put them in the GitLink.Build.Construction (and Build\Construction\Interfaces folder on disk). I think in the end that's a clearer approach.

**Comment by GeertvanHorrik** [comment]
Excellent, thanks!

**Comment by GeertvanHorrik** [comment]
btw, I completely forgot. If you want you can add yourself to the list of contributors in CONTRIBUTORS file (which I just added).

**Comment by GeertvanHorrik** [comment]
Build - 2.4.0-unstable.26 has been released to NuGet and Chocolatey

**Pull request merged**